Mathlib/Data/Multiset/Sections.lean

```diff
@@ -0,0 +1,102 @@
+/-
+Copyright (c) 2018 Johannes Hölzl. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johannes Hölzl
+
+! This file was ported from Lean 3 source module data.multiset.sections
+! leanprover-community/mathlib commit 9003f28797c0664a49e4179487267c494477d853
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Multiset.Bind
+
+/-!
+# Sections of a multiset
+-/
+
+
+namespace Multiset
+
+variable {α : Type _}
+
+section Sections
+
+/-- The sections of a multiset of multisets `s` consists of all those multisets
+which can be put in bijection with `s`, so each element is an member of the corresponding multiset.
+-/
+
+-- Porting note: `Sections` depends on `recOn` which is noncomputable.
+-- This may be removed when `Multiset.recOn` becomes computable.
+noncomputable def Sections (s : Multiset (Multiset α)) : Multiset (Multiset α) :=
+  Multiset.recOn s {0} (fun s _ c => s.bind fun a => c.map (Multiset.cons a)) fun a₀ a₁ _ pi => by
+    simp [map_bind, bind_bind a₀ a₁, cons_swap]
+#align multiset.sections Multiset.Sections
+
+@[simp]
+theorem sections_zero : Sections (0 : Multiset (Multiset α)) = {0} :=
+  rfl
+#align multiset.sections_zero Multiset.sections_zero
+
+@[simp]
+theorem sections_cons (s : Multiset (Multiset α)) (m : Multiset α) :
+    Sections (m ::ₘ s) = m.bind fun a => (Sections s).map (Multiset.cons a) :=
+  recOn_cons m s
+#align multiset.sections_cons Multiset.sections_cons
+
+theorem coe_sections :
+    ∀ l : List (List α),
+      Sections (l.map fun l : List α => (l : Multiset α) : Multiset (Multiset α)) =
+        (l.sections.map fun l : List α => (l : Multiset α) : Multiset (Multiset α))
+  | [] => rfl
+  | a :: l => by
+    simp
+    rw [← cons_coe, sections_cons, bind_map_comm, coe_sections l]
+    simp [List.sections, (· ∘ ·), List.bind]
+#align multiset.coe_sections Multiset.coe_sections
+
+@[simp]
+theorem sections_add (s t : Multiset (Multiset α)) :
+    Sections (s + t) = (Sections s).bind fun m => (Sections t).map ((· + ·) m) :=
+  Multiset.induction_on s (by simp) fun a s ih => by
+    -- Porting note: Previous code was:
+    -- simp [ih, bind_assoc, map_bind, bind_map, -add_comm]
+    --
+    -- 'add_comm' does not have [simp] attribute
```

Comment:
```suggestion
```
That's fine, the Lean code was just unoptimized. I'm sure we don't need to revisit this in the future.